ADR: Prefer TypeScript

suzubara opened the pull request:
This is an Engineering practice ADR that states a general preference for using TypeScript over JavaScript on projects. Please add feedback in comments, suggestions, or make your own commits to this branch!

jim commented on practices/appeng/adrs/0001-use-typescript.md:
```diff
@@ -0,0 +1,87 @@
+
+As consultants, we want to write understandable, maintainable code that people can use in our absence. Types provide a kind of built-in documentation for anyone looking at the code.
+
+​​TypeScript has many helpful autocomplete and snippet generation features that make development faster.
```
I don't find development with TypeScript to be faster. Certain operations such as getting a function's arguments correctly written are slightly quicker, but I spend far more time fighting with the compiler than is saved in this way. A lot of this stems from the use of third party libraries with type definitions of varying quality/completeness, but on EASi we were careful about our dependency selection and still ran into this issue.

jim commented on practices/appeng/adrs/0001-use-typescript.md:
```diff
@@ -0,0 +1,105 @@
+  to get on-boarded
+- `-` Potentially considerable learning overhead
+- `-` Typically compiled at build time, so run-time type errors (due to external data, user input, etc.) are not eliminated
+- `-` Requires changes to build system and config - need to be compiling types for local dev, before deploy, etc. Also need to manipulate the tsconfig. However, most widely used tools have clear docs or libraries that assist with this.
```
- TypeScript is being rapidly developed and upgrading to new versions can be non-trivial as tooling across the project tends to need to be upgraded collectively.

jim commented on practices/appeng/adrs/0001-use-typescript.md:
```diff
@@ -0,0 +1,105 @@
+  runtime, since type errors are captured at compile time
+- `+` Typescript has overlapping [object oriented concepts](https://levelup.gitconnected.com/typescript-object-oriented-concepts-in-a-nutshell-cb2fdeeffe6e) to a deeper level than [ES6 classes](https://www.sitepoint.com/object-oriented-javascript-deep-dive-es6-classes/) and traditional JavaScript.
+- `+` Code is clearer and easier to understand with [static typing](https://www.typescriptlang.org/docs/handbook/basic-types.html) and [interfaces](https://www.typescriptlang.org/docs/handbook/interfaces.html)
+- `+` Does not require any additional setup when used with common boilerplates such as [Create React App](https://github.com/facebook/create-react-app)
```
On EASi we used CRA and still had to spend time on configuring tooling, specifically related to linters, formatters, and pre-commit checks.

suzubara commented:
@rswerve @jim I just fleshed out the ADR with more context, and hopefully addressed your points! please let me know if you feel like something should be added or changed.

macrael commented on practices/appeng/adrs/0001-use-typescript.md:
```diff
@@ -0,0 +1,99 @@
+
+The decision to use TypeScript assumes a consensus that type safety in a programming language is better than no type safety. Below are some reasons for this:
+
+- A strong static typing system can help reduce the possibility of bugs at runtime, since type errors are captured earlier, during compilation.
```
- Types free the programmer from writing an entire class of tests. You can be sure that the objects being passed into your function have fields with the correct types, many runtime checks become unneccecary and many tests cases no longer need to be written
- The most useful type check is defining wether a variable or field is required or possibly undefined. The [billion-dollar mistake](https://en.wikipedia.org/wiki/Tony_Hoare#Apologies_and_retractions) was due to an insufficienlty expressive type system

A couple suggestions. I’m sure we can drum up more. ^

jenniferlynparsons reviewed (approved):
Yay!

rswerve commented on practices/appeng/adrs/0001-use-typescript.md:
```diff
@@ -0,0 +1,107 @@
+
+- If the project has no reason to use JavaScript, there may be no reason to use TypeScript. However, this decision does not preclude the option to use TypeScript as a server language if otherwise appropriate.
+- It can be very challenging and time consuming to introduce TypeScript to an existing JavaScript code base, especially if there are many dependencies and/or integrations that may or may not have TypeScript support. In this case it might not be feasible to spend time configuring TypeScript, although the option of introducing TypeScript incrementally (such as only to new code) should also be consi ...
+- TypeScript can have a notoriously high learning curve, and may be too costly to adopt by certain projects with tight timing and/or staffing circumstances. However, that overhead in this situation should still be weighed carefully against the payoff of using TypeScript to prevent errors and reduce the need for writing tests later on.
```
```suggestion
- TypeScript can have a notoriously high learning curve, and may be too costly to adopt by certain projects with tight timing and/or staffing circumstances. However, that overhead in this situation should still be weighed carefully against the payoff of using TypeScript to prevent errors and reduce the need for writing certain kinds of tests later on.
```

rswerve commented on practices/appeng/adrs/0001-use-typescript.md:
```diff
@@ -0,0 +1,107 @@
+## When to not implement this decision
+
+- If the project has no reason to use JavaScript, there may be no reason to use TypeScript. However, this decision does not preclude the option to use TypeScript as a server language if otherwise appropriate.
+- It can be very challenging and time consuming to introduce TypeScript to an existing JavaScript code base, especially if there are many dependencies and/or integrations that may or may not have TypeScript support. In this case it might not be feasible to spend time configuring TypeScript, although the option of introducing TypeScript incrementally (such as only to new code) should also be consi ...
```
Worth mentioning JSDoc as a possible incremental step (or compromise solution)?

rswerve replied:
☝️ genuine question, by the way, not a passive-aggressive suggestion 🙂

suzubara replied:
not at all! I am not familiar with JSDoc though, do you mind adding a suggestion?

rswerve replied:
done!

suzubara replied:
thanks!!

rswerve commented on practices/appeng/adrs/0001-use-typescript.md:
```diff
@@ -0,0 +1,107 @@
+## When to not implement this decision
+
+- If the project has no reason to use JavaScript, there may be no reason to use TypeScript. However, this decision does not preclude the option to use TypeScript as a server language if otherwise appropriate.
+- It can be very challenging and time consuming to introduce TypeScript to an existing JavaScript code base, especially if there are many dependencies and/or integrations that may or may not have TypeScript support. In this case it might not be feasible to spend time configuring TypeScript, although the option of introducing TypeScript incrementally (such as only to new code) should also be consi ...
```
```suggestion
- It can be very challenging and time consuming to introduce TypeScript to an existing JavaScript code base, especially if there are many dependencies and/or integrations that may or may not have TypeScript support. In this case it might not be feasible to spend time configuring TypeScript, although the option of introducing TypeScript incrementally (such as only to new code, or by using [JSDoc](https://www.typescriptlang.org/play#example/jsdoc-support)) should also be considered.
```

gidjin commented on practices/appeng/adrs/0001-use-typescript.md:
Does this fact have an effect on how easy it is to debug code while running? Ie if using the browser to step through some code will you only have access to the compiled JS or are there tools to map it to the original typescript? Mostly just asking for information, though might be good to include some details either way. However the advantage of type safety would out weigh that as a con imo.

suzubara replied:
@gidjin hmm good question! This is not an issue I've run into, I think because the difference between TS and the compiled JS code is not significant.. TS is essentially just adding type annotations, which get stripped out during compilation, without much change to the code itself.

gidjin reviewed (commented):
This is great! I left a question but I'm good approving once it's out of the WIP stage.
